### ISSUE TITLE
Kart model starts backwards on some arenas

### ISSUE THREAD

**Issue by stkAnon** [supertuxkart/stk-code#1548]
On some arenas (The Stadium, my temple track, possible some more addon-ons), the kart model of player two starts out facing the camera and stays that way until the kart's orientation changes, usually either through turning or collision with another kart.

This seems to happen when the start position of a kart has a transformation of -180 in Blender.

**Comment by deveee** [comment]
8da557d1537d6fb11164ad2422392f101da476f1 is the first bad commit
commit 8da557d1537d6fb11164ad2422392f101da476f1
Author: vlj <vljn@ovi.com>
Date:   Wed Sep 3 17:59:52 2014 +0200

    Use GLEW for our GL function support.

:040000 040000 7d263dac37c8863e98afa87b5c3e68e9b95981bb 2bad9df72e7b3c3e0efeb59c3173106a5a501b43 M	src

**Comment by vlj** [comment]
I don't think adding GLEW would cause this :/

**Comment by auriamg** [comment]
No, though maybe during the transition to glew some memory corruption was introduced? Or maybe some extension that was previously disabled is now enabled

**Comment by deveee** [comment]
That's true what auria says.

When I disabled GL_ARB_buffer_storage and GL_ARB_base_instance extensions in central_settings.cpp file, the bug disappeared.

GL_ARB_base_instance is the reason, but I needed to disable also GL_ARB_buffer_storage to avoid a crash.

But still hard to say where is the real reason. Probably somewhere between isARBBaseInstanceUsable() functions...

**Comment by deveee** [comment]
@vlj not sure if this will help, but it works for me when I disable these lines:
```
diff --git a/src/graphics/stkscenemanager.cpp b/src/graphics/stkscenemanager.cpp
index 97a48d7..0300419 100644
--- a/src/graphics/stkscenemanager.cpp
+++ b/src/graphics/stkscenemanager.cpp
@@ -301,9 +301,9 @@ handleSTKCommon(scene::ISceneNode *Node, std::vector<scene::ISceneNode *> *Immed
                     if (node->glow())
                         MeshForGlowPass[mesh->mb].emplace_back(mesh, Node);
 
-                    if (Mat != Material::SHADERTYPE_SPLATTING && mesh->TextureMatrix.isIdentity())
-                        MeshForSolidPass[Mat][mesh->mb].emplace_back(mesh, Node);
-                    else
+                    //~ if (Mat != Material::SHADERTYPE_SPLATTING && mesh->TextureMatrix.isIdentity())
+                        //~ MeshForSolidPass[Mat][mesh->mb].emplace_back(mesh, Node);
+                    //~ else
                     {
                         core::matrix4 ModelMatrix = Node->getAbsoluteTransformation(), InvModelMatrix;
                         ModelMatrix.getInverse(InvModelMatrix);
```

**Comment by vlj** [comment]
Ok so this means that matrix4::getOrientation method is buggy in Irrlicht. No idea what change between race track and soccer track though.

**Comment by hiker** [comment]
@deveee can you verify that indeed mesh->TextureMatrix.isIdentity() is incorrect? At least I assume that's the explanation: a non-identical matrix being flagged as identical; then when a minor change happens, identity is reset, and the correct (but more expensive) code is used.
If this is the reason, and we can't find the real reason, we should apply the patch as above - I would suspect that having an identity matrix there is the exception.

**Comment by deveee** [comment]
Really hard to say what's wrong. isIdentity() function seems to be ok. I uncommented code in this function which does some additional checks, but it didn't help.

I also noticed that modification from my previous comment solves kart flickering, but the kart's shadow still flickers. So the reason must be different than isIdentity() function.

**Comment by hiker** [comment]
Hmm - if I understand you correctly you removed this condition (and the code that would be executed), ```if (Mat != Material::SHADERTYPE_SPLATTING && mesh->TextureMatrix.isIdentity())```
forcing always the slower way (inverting the matrix).
So if this is in case, then either the Mat!=SPLATTING is incorrect, or isIdentity is .. not doing what we expect it to do. Could you verify which of the two conditions is wrong? Worst case, just log the whole matrix, and output of isIdentity() (if it's the matrix). 
If it's the matrix, it might be that the epsilon used for the floating point comparison is too big, i.e.the matrix appears to be identity because of the epsilon, but is de-facto not. A bit surprising that this should make that much of a difference though. Could also be a compiler problem or so.

You can also try to use isIdentity_integer_base(), which does not have an epsilon, it tests of exactly 0 and 1.
Note that I can't reproduce since I can't go (with my hd3000) into that path of the code

**Comment by vlj** [comment]
actually the isIdentity is the rule, not the exception. this is the texture matrix, not the modelmatrix

**Comment by vlj** [comment]
I suspect something similar to what's happening with the adiummy wrong rotation interpolation bug that was fixed in the exporter : we try to get rotation from a matrix and for whatever reason we hits one corner case. Probably the fact that the mesh is skinned and thus uses quaternion to matrix doesn't help.

**Comment by vlj** [comment]
I found the issue : the matrix for the kart mesh is 
(-1       0             0       0)
(-0       1            -0       0)
(-0      -0            -1       0)
(-4.53 0.002828 35.07 1)

Rotation is 180 360 180
Scale is -1 1 -1

Looks like irrlicht doesnt want to decide between a negative scale and a 180° rotation and thus the decomposition is wrong.
I fixed it by forcing scale to be positive but it may introduces issue in others places.

**Comment by vlj** [comment]
No fix until further notice.

**Comment by deveee** [comment]
Not sure what happens, I just restore vlj's comment because it ma be useful...

```
I found the issue : the matrix for the kart mesh is 
(-1       0             0       0)
(-0       1            -0       0)
(-0      -0            -1       0)
(-4.53 0.002828 35.07 1)

Rotation is 180 360 180
Scale is -1 1 -1

Looks like irrlicht doesnt want to decide between a negative scale and a 180° rotation and thus the decomposition is wrong.
I fixed it by forcing scale to be positive but it may introduces issue in others places.
```

**Comment by hiker** [comment]
I had a look at the getRotationDegrees() appears to be correct, but getScale returns incorrect (negative) values. getRotationDegrees() takes this into account:
```
		// we need to check for negative scale on to axes, which would bring up wrong results
		if (scale.Y<0 && scale.Z<0)
		{
			scale.Y =-scale.Y;
			scale.Z =-scale.Z;
		}
```
etc.
So, three different solutions to try/verify:
1. modify getScale of irrlicht's matrix.h to always return positive scaling (it already does that in case of non diagonal matrices - so it's only fabsf() in the case that the 6-non-diagonal matrix elements are zero).
2. Modify getScale() to include the sign changes implemented in getRotationDegrees() (if two scales are negative, switch them both).
3. If any of the above should have side-effects, we might only need to modify the result of getScale() in stkscenemanager.
@deveee can you give that a try?

**Comment by hiker** [comment]
Actually, solution 2) is probably the one to try first, since it should still allow us to do mirroring on a single axis (mirroring on two axes will always appears as a rotation)

**Comment by deveee** [comment]
I will look at this.
Btw. when I always do full calculation in getScale() in matrix4.h, I 
don't see this bug anymore.

**Comment by hiker** [comment]
Correct, the full calculation does never returns negative values. Since afaik we don't use negative scales (caused by a 180 degree heading rotation being stored as 180 degree pitch followed by 180 degree roll, which is indeed identical), it should be fine to just add fabs around the values. If we should ever need negative scaling, removing two negative scales (which are just a rotation and two positive scales) should all work fine.

**Comment by deveee** [comment]
I added checks for negative values in getScale() function. They are exactly the same as in getRotationDegrees().

```
diff --git a/lib/irrlicht/include/matrix4.h b/lib/irrlicht/include/matrix4.h
index f4af53d..cd546be 100644
--- a/lib/irrlicht/include/matrix4.h
+++ b/lib/irrlicht/include/matrix4.h
@@ -868,7 +868,22 @@ namespace core
 		if(core::iszero(M[1]) && core::iszero(M[2]) &&
 			core::iszero(M[4]) && core::iszero(M[6]) &&
 			core::iszero(M[8]) && core::iszero(M[9]))
+		{
+			if (M[5]<0 && M[10]<0)
+			{
+				return vector3d<T>(M[0], -M[5], -M[10]);
+			}
+			else if (M[0]<0 && M[10]<0)
+			{
+				return vector3d<T>(-M[0], M[5], -M[10]);
+			}
+			else if (M[0]<0 && M[5]<0)
+			{
+				return vector3d<T>(-M[0], -M[5], M[10]);
+			}
+			
 			return vector3d<T>(M[0], M[5], M[10]);
+		}
 
 		// We have to do the full calculation.
 		return vector3d<T>(sqrtf(M[0] * M[0] + M[1] * M[1] + M[2] * M[2]),

```

After this modification everything is fine. This bug disappeared and after quick test I didn't notice any side-effects.

But... I also noticed that core::iszero() function can cause problems with negative values. If value is for example -0.00001 it returns true. In full calculation this negative value will change sign of result. But we just ignore it.

I made another patch:
```
diff --git a/lib/irrlicht/include/matrix4.h b/lib/irrlicht/include/matrix4.h
index f4af53d..bdac1b0 100644
--- a/lib/irrlicht/include/matrix4.h
+++ b/lib/irrlicht/include/matrix4.h
@@ -865,9 +865,9 @@ namespace core
 
 		// Deal with the 0 rotation case first
 		// Prior to Irrlicht 1.6, we always returned this value.
-		if(core::iszero(M[1]) && core::iszero(M[2]) &&
-			core::iszero(M[4]) && core::iszero(M[6]) &&
-			core::iszero(M[8]) && core::iszero(M[9]))
+		if(M[1] == 0 && M[2] == 0 &&
+			M[4] == 0 && M[6] == 0 &&
+			M[8] == 0 && M[9] == 0)
 			return vector3d<T>(M[0], M[5], M[10]);
 
 		// We have to do the full calculation.

```

It also works fine for me. But it will do full calculations more often (values are not rounded anymore).

@hiker what do you think about it?

**Comment by deveee** [comment]
I think that second solution is fine. It still doesn't do full calculation in most cases. Only in case when some of M[1], M[2], M[4], M[6], M[8], M[9] have negative values. I will test more carefully if it doesn't have any side effects.

**Comment by deveee** [comment]
After more testing, I generally observe three situations:
- M[1], M[2], M[4], M[6], M[8], M[9] are precisely equal to 0 and we don't need to do full calculation
- M[1], M[2], M[4], M[6], M[8], M[9] have rather high values and we have to do full calculation
- One specific case in soccer arena where M[1], M[2], M[4], M[6], M[8], M[9] have very low and negative values (-0.000001 or so).

I don't observe situations where we have very low and positive values. So I think that core::iszero() function is useless. And when we change it to M[1]==0 etc. we just handle this one specific case by doing full calculation.

I hope that it's nothing wrong in my deduction...

But probably the real reason of these very low negative values is still in other place :P If you don't observe it on Windows, maybe it is different rounding, different number precision, optimization or so...

**Comment by deveee** [comment]
Fixed in 0722df79da9e273bf1a83aef7446045ea9cd24ae.
Feel free to revert it if you will notice some side-effects.
